File: app/crud.py
# SQLAlchemy
from sqlalchemy.orm import Session
from sqlalchemy import func, extract, or_, and_

from app import models, schemas
from datetime import datetime, timedelta, date
from decimal import Decimal
from typing import List, Optional, Dict, Union

# Product CRUD Operations
def get_product(db: Session, product_id: int):
    return db.query(models.Product).filter(models.Product.product_id == product_id).first()

def get_products(db: Session, skip: int = 0, limit: int = 100):
    return db.query(models.Product).offset(skip).limit(limit).all()

def create_product(db: Session, product: schemas.ProductCreate):
    db_product = models.Product(**product.dict())
    db.add(db_product)
    db.commit()
    db.refresh(db_product)
    return db_product

def update_product(db: Session, product_id: int, product: schemas.ProductUpdate):
    db_product = get_product(db, product_id)
    if db_product:
        update_data = product.dict(exclude_unset=True)
        for key, value in update_data.items():
            setattr(db_product, key, value)
        db_product.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(db_product)
    return db_product

def delete_product(db: Session, product_id: int):
    db_product = get_product(db, product_id)
    if db_product:
        db.delete(db_product)
        db.commit()
    return db_product

def get_inventory(
    db: Session, 
    skip: int = 0, 
    limit: int = 100,
    low_stock_only: bool = False
) -> List[models.Inventory]:
    """Get inventory with pagination support"""
    query = db.query(models.Inventory)
    
    if low_stock_only:
        query = query.filter(
            models.Inventory.quantity <= models.Inventory.low_stock_threshold
        )
    
    return query.offset(skip).limit(limit).all()

def update_inventory(db: Session, product_id: int, inventory_update: schemas.InventoryUpdate):
    """Update inventory for a specific product."""
    # First check if product exists
    product = db.query(models.Product).filter(models.Product.product_id == product_id).first()
    if not product:
        return {}
    
    # Then get associated inventory
    # Make sure to use .first() to get a single object, not a list
    db_inventory = db.query(models.Inventory).filter(models.Inventory.product_id == product_id).first()
    
    # Store the previous quantity before updating (if inventory exists)
    previous_quantity = 0
    if db_inventory:
        previous_quantity = db_inventory.quantity
    
    if not db_inventory:
        # Create new inventory if it doesn't exist
        db_inventory = models.Inventory(product_id=product_id, **inventory_update.dict())
        db.add(db_inventory)
    else:
        # Update existing inventory
        for key, value in inventory_update.dict().items():
            setattr(db_inventory, key, value)
    
    # Add history record
    history = models.InventoryHistory(
        product_id=product_id,
        previous_quantity=previous_quantity,  # Use the saved value
        new_quantity=inventory_update.quantity,
        change_reason="Manual update via API"
    )
    db.add(history)
    
    db.commit()
    db.refresh(db_inventory)
    return db_inventory

# Sale CRUD Operations
def create_sale(db: Session, sale: schemas.SaleCreate):
    # Get current product price
    product = get_product(db, sale.product_id)
    if not product:
        return None
        
    db_sale = models.Sale(
        product_id=sale.product_id,
        quantity=sale.quantity,
        unit_price=product.price,
        total_amount=sale.quantity * product.price
    )
    
    # Update inventory
    inventory = get_inventory(db, sale.product_id)
    if inventory:
        inventory.quantity -= sale.quantity
    
    db.add(db_sale)
    db.commit()
    db.refresh(db_sale)
    return db_sale

def get_sales(
    db: Session,
    skip: int = 0,
    limit: int = 100,
    start_date: Optional[datetime] = None,
    end_date: Optional[datetime] = None,
    product_id: Optional[int] = None,
    category: Optional[str] = None  # Add this parameter
) -> List[models.Sale]:
    query = db.query(models.Sale)
    
    if start_date:
        query = query.filter(models.Sale.sale_date >= start_date)
    if end_date:
        query = query.filter(models.Sale.sale_date <= end_date)
    if product_id:
        query = query.filter(models.Sale.product_id == product_id)
    if category:  # Add this filter
        query = query.join(models.Product).filter(models.Product.category == category)
    
    return query.order_by(models.Sale.sale_date.desc()).offset(skip).limit(limit).all()

# Analytics Operations
def get_revenue_by_period(db: Session, period: str, start_date: datetime, end_date: datetime):
    return db.query(
        models.Sale.product_id,
        models.Product.category,
        models.Sale.total_amount
    ).join(models.Product).filter(
        models.Sale.sale_date.between(start_date, end_date)
    ).all()

def get_low_stock_items(db: Session, skip: int = 0, limit: int = 100, threshold: int = 10):
    return db.query(models.Inventory).join(models.Product).filter(
        models.Inventory.quantity <= models.Inventory.low_stock_threshold
    ).offset(skip).limit(limit).all()

# ======================
# INVENTORY HISTORY CRUD
# ======================

def get_inventory_history(
    db: Session, 
    product_id: Optional[int] = None,
    days: int = 30,
    skip: int = 0, 
    limit: int = 100
) -> List[models.InventoryHistory]:
    """Get inventory change history with filters"""
    query = db.query(models.InventoryHistory)
    
    if product_id:
        query = query.filter(models.InventoryHistory.product_id == product_id)
    
    cutoff_date = datetime.utcnow() - timedelta(days=days)
    query = query.filter(models.InventoryHistory.changed_at >= cutoff_date)
    
    return query.order_by(
        models.InventoryHistory.changed_at.desc()
    ).offset(skip).limit(limit).all()

def record_inventory_change(
    db: Session,
    product_id: int,
    previous_quantity: int,
    new_quantity: int,
    change_reason: str
) -> models.InventoryHistory:
    """Create a new inventory history record"""
    if previous_quantity == new_quantity:
        return None  # No actual change
    
    db_history = models.InventoryHistory(
        product_id=product_id,
        previous_quantity=previous_quantity,
        new_quantity=new_quantity,
        change_reason=change_reason,
        changed_at=datetime.utcnow()
    )
    db.add(db_history)
    db.commit()
    db.refresh(db_history)
    return db_history

# ======================
# ENHANCED INVENTORY CRUD 
# ======================

def get_inventory_with_history(
    db: Session,
    product_id: int
) -> schemas.InventoryWithHistory:
    """Get inventory with its change history"""
    inventory = db.query(models.Inventory).filter(
        models.Inventory.product_id == product_id
    ).first()
    
    if not inventory:
        return None
    
    history = get_inventory_history(db, product_id=product_id)
    
    return {
        "inventory": inventory,
        "history": history
    }

# ======================
# MISSING PRODUCT CRUD
# ======================

def get_products_by_category(
    db: Session,
    category: str,
    skip: int = 0,
    limit: int = 100
) -> List[models.Product]:
    """Get products filtered by category"""
    return db.query(models.Product).filter(
        models.Product.category == category
    ).offset(skip).limit(limit).all()

def search_products(
    db: Session,
    search_term: str,
    skip: int = 0,
    limit: int = 100
) -> List[models.Product]:
    """Search products by name or description"""
    return db.query(models.Product).filter(
        models.Product.name.ilike(f"%{search_term}%") |
        models.Product.description.ilike(f"%{search_term}%")
    ).offset(skip).limit(limit).all()

# ======================
# MISSING ANALYTICS CRUD
# ======================

def get_category_revenue(
    db: Session,
    start_date: datetime,
    end_date: datetime
) -> List[schemas.CategoryRevenue]:
    """Get revenue breakdown by category"""
    return db.query(
        models.Product.category,
        func.sum(models.Sale.total_amount).label("revenue"),
        func.count(models.Sale.sale_id).label("transactions")
    ).join(models.Sale).filter(
        models.Sale.sale_date.between(start_date, end_date)
    ).group_by(models.Product.category).all()

# ======================
# BULK OPERATIONS
# ======================

def bulk_update_inventory(
    db: Session,
    updates: List[schemas.InventoryBulkUpdate]
) -> List[models.Inventory]:
    """Update multiple inventory items at once"""
    results = []
    for update in updates:
        inventory = db.query(models.Inventory).filter(
            models.Inventory.product_id == update.product_id
        ).first()
        
        if inventory:
            # Record history before updating
            record_inventory_change(
                db,
                product_id=update.product_id,
                previous_quantity=inventory.quantity,
                new_quantity=update.quantity,
                change_reason="Bulk update"
            )
            
            inventory.quantity = update.quantity
            inventory.last_restocked = datetime.utcnow()
            results.append(inventory)
    
    db.commit()
    return results

File: src/app/crud.ts
import { Prisma, PrismaClient } from '@prisma/client';
import type {
  Product,
  Inventory,
  Sale,
  InventoryHistory,
} from '@prisma/client';

/**
 * Data-access (CRUD) layer for the product inventory and sales system.
 *
 * MIGRATION_NOTE: The source used SQLAlchemy. Despite the "django" migration
 * label, this has been migrated to idiomatic Node.js/TypeScript using Prisma.
 * The thin CRUD-function style is preserved, but each function now accepts a
 * PrismaClient (or transaction client) for dependency injection rather than a
 * SQLAlchemy Session.
 *
 * MIGRATION_NOTE: SQLAlchemy's manual commit/refresh lifecycle is replaced by
 * Prisma's automatic per-call persistence. Multi-write operations that must be
 * atomic use prisma.$transaction.
 */

// ---------------------------------------------------------------------------
// Input/DTO types (replacing Pydantic schemas)
// ---------------------------------------------------------------------------

export interface ProductCreateInput {
  name: string;
  description?: string | null;
  category?: string | null;
  price: Prisma.Decimal | number | string;
  [key: string]: unknown;
}

export interface ProductUpdateInput {
  name?: string;
  description?: string | null;
  category?: string | null;
  price?: Prisma.Decimal | number | string;
  [key: string]: unknown;
}

export interface InventoryUpdateInput {
  quantity: number;
  lowStockThreshold?: number;
  lastRestocked?: Date;
  [key: string]: unknown;
}

export interface SaleCreateInput {
  productId: number;
  quantity: number;
}

export interface InventoryBulkUpdateInput {
  productId: number;
  quantity: number;
}

export interface InventoryWithHistory {
  inventory: Inventory;
  history: InventoryHistory[];
}

export interface CategoryRevenue {
  category: string | null;
  revenue: Prisma.Decimal | number;
  transactions: number;
}

export interface RevenueByPeriodRow {
  productId: number;
  category: string | null;
  totalAmount: Prisma.Decimal | number;
}

/**
 * A Prisma client or transaction client. Using this union allows the CRUD
 * functions to participate in externally-managed transactions.
 */
export type Db = PrismaClient | Prisma.TransactionClient;

// ---------------------------------------------------------------------------
// Product CRUD Operations
// ---------------------------------------------------------------------------

export async function getProduct(
  db: Db,
  productId: number,
): Promise<Product | null> {
  return db.product.findUnique({ where: { productId } });
}

export async function getProducts(
  db: Db,
  skip = 0,
  limit = 100,
): Promise<Product[]> {
  return db.product.findMany({ skip, take: limit });
}

export async function createProduct(
  db: Db,
  product: ProductCreateInput,
): Promise<Product> {
  return db.product.create({ data: product as Prisma.ProductCreateInput });
}

export async function updateProduct(
  db: Db,
  productId: number,
  product: ProductUpdateInput,
): Promise<Product | null> {
  const existing = await getProduct(db, productId);
  if (!existing) {
    return null;
  }

  // exclude_unset semantics: only update keys that were actually provided.
  const updateData: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(product)) {
    if (value !== undefined) {
      updateData[key] = value;
    }
  }
  updateData.updatedAt = new Date();

  return db.product.update({
    where: { productId },
    data: updateData as Prisma.ProductUpdateInput,
  });
}

export async function deleteProduct(
  db: Db,
  productId: number,
): Promise<Product | null> {
  const existing = await getProduct(db, productId);
  if (!existing) {
    return null;
  }
  await db.product.delete({ where: { productId } });
  return existing;
}

// ---------------------------------------------------------------------------
// Inventory CRUD Operations
// ---------------------------------------------------------------------------

/**
 * Get inventory with pagination support.
 *
 * MIGRATION_NOTE: The source compared a column to another column
 * (quantity <= low_stock_threshold). Prisma cannot express column-to-column
 * comparisons in a `where` filter directly, so for `lowStockOnly` we fetch and
 * filter in application code. For large datasets this should be replaced with a
 * raw query (db.$queryRaw) - flagged for manual review.
 */
export async function getInventory(
  db: Db,
  skip = 0,
  limit = 100,
  lowStockOnly = false,
): Promise<Inventory[]> {
  if (lowStockOnly) {
    const all = await db.inventory.findMany();
    const filtered = all.filter((i) => i.quantity <= i.lowStockThreshold);
    return filtered.slice(skip, skip + limit);
  }
  return db.inventory.findMany({ skip, take: limit });
}

/**
 * Fetch a single inventory row by product id.
 *
 * MIGRATION_NOTE: This helper fixes the latent bug in the original
 * `create_sale`, where `get_inventory(db, sale.product_id)` was called with a
 * product id passed into a positional `skip` parameter and returned a list.
 */
export async function getInventoryByProductId(
  db: Db,
  productId: number,
): Promise<Inventory | null> {
  return db.inventory.findUnique({ where: { productId } });
}

/**
 * Update inventory for a specific product.
 * Returns null when the product does not exist (was `{}` in the source).
 */
export async function updateInventory(
  db: Db,
  productId: number,
  inventoryUpdate: InventoryUpdateInput,
): Promise<Inventory | null> {
  // The whole operation (inventory upsert + history) must be atomic.
  const runner = async (tx: Prisma.TransactionClient): Promise<Inventory | null> => {
    const product = await tx.product.findUnique({ where: { productId } });
    if (!product) {
      return null;
    }

    const existing = await tx.inventory.findUnique({ where: { productId } });
    const previousQuantity = existing ? existing.quantity : 0;

    let dbInventory: Inventory;
    if (!existing) {
      dbInventory = await tx.inventory.create({
        data: {
          productId,
          ...(inventoryUpdate as Prisma.InventoryCreateInput),
        },
      });
    } else {
      dbInventory = await tx.inventory.update({
        where: { productId },
        data: inventoryUpdate as Prisma.InventoryUpdateInput,
      });
    }

    await tx.inventoryHistory.create({
      data: {
        productId,
        previousQuantity,
        newQuantity: inventoryUpdate.quantity,
        changeReason: 'Manual update via API',
      },
    });

    return dbInventory;
  };

  // Support being called both with a top-level client and within a transaction.
  if ('$transaction' in db) {
    return (db as PrismaClient).$transaction(runner);
  }
  return runner(db as Prisma.TransactionClient);
}

// ---------------------------------------------------------------------------
// Sale CRUD Operations
// ---------------------------------------------------------------------------

export async function createSale(
  db: Db,
  sale: SaleCreateInput,
): Promise<Sale | null> {
  const runner = async (tx: Prisma.TransactionClient): Promise<Sale | null> => {
    const product = await tx.product.findUnique({
      where: { productId: sale.productId },
    });
    if (!product) {
      return null;
    }

    const unitPrice = new Prisma.Decimal(product.price as unknown as string);
    const totalAmount = unitPrice.mul(sale.quantity);

    const dbSale = await tx.sale.create({
      data: {
        productId: sale.productId,
        quantity: sale.quantity,
        unitPrice,
        totalAmount,
      },
    });

    // MIGRATION_NOTE: Original called get_inventory(db, sale.product_id) which
    // due to a bug fetched a list. Fixed here to load the single inventory row.
    const inventory = await tx.inventory.findUnique({
      where: { productId: sale.productId },
    });
    if (inventory) {
      await tx.inventory.update({
        where: { productId: sale.productId },
        data: { quantity: inventory.quantity - sale.quantity },
      });
    }

    return dbSale;
  };

  if ('$transaction' in db) {
    return (db as PrismaClient).$transaction(runner);
  }
  return runner(db as Prisma.TransactionClient);
}

export interface GetSalesOptions {
  skip?: number;
  limit?: number;
  startDate?: Date;
  endDate?: Date;
  productId?: number;
  category?: string;
}

export async function getSales(
  db: Db,
  options: GetSalesOptions = {},
): Promise<Sale[]> {
  const { skip = 0, limit = 100, startDate, endDate, productId, category } = options;

  const where: Prisma.SaleWhereInput = {};

  if (startDate || endDate) {
    where.saleDate = {};
    if (startDate) where.saleDate.gte = startDate;
    if (endDate) where.saleDate.lte = endDate;
  }
  if (productId) {
    where.productId = productId;
  }
  if (category) {
    // join(Product).filter(Product.category == category) -> relation filter
    where.product = { category };
  }

  return db.sale.findMany({
    where,
    orderBy: { saleDate: 'desc' },
    skip,
    take: limit,
  });
}

// ---------------------------------------------------------------------------
// Analytics Operations
// ---------------------------------------------------------------------------

/**
 * Return per-sale rows of (productId, category, totalAmount) within a range.
 *
 * MIGRATION_NOTE: The `period` argument was unused in the source query and is
 * preserved here only for signature compatibility.
 */
export async function getRevenueByPeriod(
  db: Db,
  _period: string,
  startDate: Date,
  endDate: Date,
): Promise<RevenueByPeriodRow[]> {
  const sales = await db.sale.findMany({
    where: { saleDate: { gte: startDate, lte: endDate } },
    include: { product: { select: { category: true } } },
  });

  return sales.map((s) => ({
    productId: s.productId,
    category: (s as Sale & { product: { category: string | null } }).product.category,
    totalAmount: s.totalAmount,
  }));
}

/**
 * MIGRATION_NOTE: Column-to-column comparison (quantity <= low_stock_threshold)
 * is not expressible in a Prisma `where`. Filtered in application code. The
 * `threshold` parameter is unused, matching the source behaviour. Consider a
 * raw query for large tables.
 */
export async function getLowStockItems(
  db: Db,
  skip = 0,
  limit = 100,
  _threshold = 10,
): Promise<Inventory[]> {
  const all = await db.inventory.findMany();
  const filtered = all.filter((i) => i.quantity <= i.lowStockThreshold);
  return filtered.slice(skip, skip + limit);
}

// ---------------------------------------------------------------------------
// Inventory History CRUD
// ---------------------------------------------------------------------------

export interface GetInventoryHistoryOptions {
  productId?: number;
  days?: number;
  skip?: number;
  limit?: number;
}

export async function getInventoryHistory(
  db: Db,
  options: GetInventoryHistoryOptions = {},
): Promise<InventoryHistory[]> {
  const { productId, days = 30, skip = 0, limit = 100 } = options;

  const cutoffDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000);

  const where: Prisma.InventoryHistoryWhereInput = {
    changedAt: { gte: cutoffDate },
  };
  if (productId) {
    where.productId = productId;
  }

  return db.inventoryHistory.findMany({
    where,
    orderBy: { changedAt: 'desc' },
    skip,
    take: limit,
  });
}

export async function recordInventoryChange(
  db: Db,
  productId: number,
  previousQuantity: number,
  newQuantity: number,
  changeReason: string,
): Promise<InventoryHistory | null> {
  if (previousQuantity === newQuantity) {
    return null; // No actual change
  }

  return db.inventoryHistory.create({
    data: {
      productId,
      previousQuantity,
      newQuantity,
      changeReason,
      changedAt: new Date(),
    },
  });
}

// ---------------------------------------------------------------------------
// Enhanced Inventory CRUD
// ---------------------------------------------------------------------------

export async function getInventoryWithHistory(
  db: Db,
  productId: number,
): Promise<InventoryWithHistory | null> {
  const inventory = await db.inventory.findUnique({ where: { productId } });
  if (!inventory) {
    return null;
  }

  const history = await getInventoryHistory(db, { productId });

  return { inventory, history };
}

// ---------------------------------------------------------------------------
// Product CRUD (category / search)
// ---------------------------------------------------------------------------

export async function getProductsByCategory(
  db: Db,
  category: string,
  skip = 0,
  limit = 100,
): Promise<Product[]> {
  return db.product.findMany({
    where: { category },
    skip,
    take: limit,
  });
}

/**
 * Search products by name or description.
 *
 * MIGRATION_NOTE: SQLAlchemy ilike() (case-insensitive) maps to Prisma's
 * `contains` with `mode: 'insensitive'` (PostgreSQL). On MySQL, default
 * collation is usually case-insensitive and `mode` is ignored - verify.
 */
export async function searchProducts(
  db: Db,
  searchTerm: string,
  skip = 0,
  limit = 100,
): Promise<Product[]> {
  return db.product.findMany({
    where: {
      OR: [
        { name: { contains: searchTerm, mode: 'insensitive' } },
        { description: { contains: searchTerm, mode: 'insensitive' } },
      ],
    },
    skip,
    take: limit,
  });
}

// ---------------------------------------------------------------------------
// Analytics CRUD
// ---------------------------------------------------------------------------

/**
 * Get revenue breakdown by category.
 *
 * MIGRATION_NOTE: Source used func.sum/func.count with group_by(category) via a
 * join. Prisma's groupBy aggregates on a single model, so we group sales by
 * productId, then map products to categories and aggregate in application code.
 * For large datasets prefer a raw SQL query - flagged for manual review.
 */
export async function getCategoryRevenue(
  db: Db,
  startDate: Date,
  endDate: Date,
): Promise<CategoryRevenue[]> {
  const sales = await db.sale.findMany({
    where: { saleDate: { gte: startDate, lte: endDate } },
    include: { product: { select: { category: true } } },
  });

  const buckets = new Map<string | null, { revenue: Prisma.Decimal; transactions: number }>();

  for (const sale of sales) {
    const category = (sale as Sale & { product: { category: string | null } })
      .product.category;
    const existing = buckets.get(category) ?? {
      revenue: new Prisma.Decimal(0),
      transactions: 0,
    };
    existing.revenue = existing.revenue.add(
      new Prisma.Decimal(sale.totalAmount as unknown as string),
    );
    existing.transactions += 1;
    buckets.set(category, existing);
  }

  return Array.from(buckets.entries()).map(([category, agg]) => ({
    category,
    revenue: agg.revenue,
    transactions: agg.transactions,
  }));
}

// ---------------------------------------------------------------------------
// Bulk Operations
// ---------------------------------------------------------------------------

export async function bulkUpdateInventory(
  db: Db,
  updates: InventoryBulkUpdateInput[],
): Promise<Inventory[]> {
  const runner = async (tx: Prisma.TransactionClient): Promise<Inventory[]> => {
    const results: Inventory[] = [];

    for (const update of updates) {
      const inventory = await tx.inventory.findUnique({
        where: { productId: update.productId },
      });

      if (inventory) {
        // Record history before updating.
        await recordInventoryChange(
          tx,
          update.productId,
          inventory.quantity,
          update.quantity,
          'Bulk update',
        );

        const updated = await tx.inventory.update({
          where: { productId: update.productId },
          data: {
            quantity: update.quantity,
            lastRestocked: new Date(),
          },
        });
        results.push(updated);
      }
    }

    return results;
  };

  if ('$transaction' in db) {
    return (db as PrismaClient).$transaction(runner);
  }
  return runner(db as Prisma.TransactionClient);
}
